File: python/tefeta.py
#!/usr/bin/env python
import sys

down = '|'
side = '-'
wall = '*'

with open(sys.argv[1], 'r') as file:
    map = file.read().splitlines()
    pathed_map = []
    entry = []
    exit = []

    # setting start and end coordinates
    for i, line in enumerate(map):
        if i == 0:
            size = line.split('x', 1)
            width = int(size[0])
            height = int(size[1])
        else:
            for char_i, c in enumerate(line):
                if (c == '1'):
                    entry = [i, char_i]
                if (c == '2'):
                    exit = [i, char_i]

    # defining the path to reach end
    for line_i in range(entry[0], exit[0]):
        line = list(map[line_i])
        for cursor_i, cursor in enumerate(range(exit[1], entry[1], 1 if entry[1] > exit[1] else -1)):
            if line[cursor] == ' ':
                line[cursor] = side
        if line_i == height - 1:
            pathed_map.append(line[cursor])

    print(line)

    print('start :', entry)
    print('end :', exit)

File: tefeta.js
#!/usr/bin/env node
const fs = require('fs')
const os = require('os')

let map = fs.readFileSync(process.argv[2], 'utf8').split(os.platform() === 'win32' ? '\r\n' : '\n')
let path = '.'
let wall = '*'
let clean = ' '
let width
let height
let entry
let exit

// setup entry and exit coordinates
map.map((line, i) => {
  if (i === 0) {
    const size = line.split('x')
    width = size[0]
    height = size[1]
  }
  if (line.indexOf(1) > -1) entry = [i, line.indexOf(1)]
  if (line.indexOf(2) > -1) exit = [i, line.indexOf(2)]
})

if (!entry) throw 'Error: no entry point found'
if (!exit) throw 'Error: no exit point found'

// write path in lines
let vDir = entry[0] < exit[0]
let hDir = true
let cursor = entry[1]
let lineExit = null

for (let i = entry[0]; vDir ? i <= exit[0] : i >= exit[0]; vDir ? i++ : i--) {
  // search for next line possible exit
  if (!lineExit) {
    for (let j = cursor; j < width; j++)
      if (
        !lineExit &&
        (map[i][j] === clean || (map[i][j] === '1' && map[vDir ? i+1 : i-1][j] === clean)) &&
        (map[vDir ? i+1 : i-1][j] === clean || map[vDir ? i+1 : i-1][j] === '2'))
        lineExit = j
  }

  if (!lineExit) {
    for (let k = cursor; k > 0; k--)
      if (!lineExit && map[i][k] === clean && (map[vDir ? i+1 : i-1][k] === clean || map[vDir ? i+1 : i-1][k] === '2'))
        lineExit = k
  }

  // draw path in line
  if ((vDir && i >= entry[0] && i <= exit[0]) || (!vDir && i <= entry[0] && i >= exit[0]) && lineExit) {
    let newLine = ''
    hDir = cursor < lineExit

    for (let l = 0; l < width; l++) {
      if (l === lineExit && map[i][l] === clean) newLine += path
      else {
        // when ** cursor --- lineExit **
        if (hDir && l >= cursor && l <= lineExit && map[i][l] === clean) newLine += path
        // when ** lineExit --- cursor **
        else if (!hDir && l <= cursor && l >= lineExit && map[i][l] === clean) newLine += path
        else newLine += map[i][l]
      }
    }
    map[i] = newLine //+ `  <--  ${map[i]} (hDir: ${hDir}, cursor: ${cursor}, lineExit: ${lineExit})`
  } else console.log(`Error: no exit found for line ${i}`)
  cursor = lineExit
  lineExit = null
}

map.map((line, i) => console.log(line))
